File: firehose_listener.swift
#!/usr/bin/env swift

import Foundation

// Add the Petrel package directory to the search path
#if canImport(FoundationNetworking)
    import FoundationNetworking
#endif

// MARK: - Main Firehose Listener

@main
struct FirehoseListener {
    static func main() async {
        print("🔥 Petrel Firehose Listener")
        print("===========================")
        print("Connecting to Bluesky firehose...")
        print("Press Ctrl+C to stop\n")

        do {
            try await runFirehose()
        } catch {
            print("❌ Fatal error: \(error)")
            exit(1)
        }
    }

    static func runFirehose() async throws {
        // Create WebSocket connection to the firehose
        guard let url = URL(string: "wss://bsky.network/xrpc/com.atproto.sync.subscribeRepos") else {
            throw NSError(domain: "Invalid URL", code: -1)
        }

        let session = URLSession(configuration: .default)
        let webSocketTask = session.webSocketTask(with: url)
        webSocketTask.resume()

        print("✅ Connected to firehose!\n")

        var eventCount = 0
        var commitCount = 0
        var identityCount = 0
        var accountCount = 0
        var postCount = 0
        var likeCount = 0
        var followCount = 0
        var repostCount = 0
        var profileCount = 0

        let startTime = Date()

        // Listen for messages
        while true {
            do {
                let message = try await webSocketTask.receive()

                switch message {
                case .data(let data):
                    eventCount += 1

                    // Try to decode the message type from CBOR
                    if let messageInfo = try? decodeFirehoseMessage(data) {
                        switch messageInfo.type {
                        case "#commit":
                            commitCount += 1
                            if let ops = messageInfo.ops {
                                for (action, path) in ops {
                                    processOperation(
                                        action: action,
                                        path: path,
                                        postCount: &postCount,
                                        likeCount: &likeCount,
                                        followCount: &followCount,
                                        repostCount: &repostCount,
                                        profileCount: &profileCount,
                                        eventCount: eventCount
                                    )
                                }
                            }

                        case "#identity":
                            identityCount += 1
                            if identityCount % 10 == 0 || eventCount < 50 {
                                print("🆔 Identity update event")
                            }

                        case "#account":
                            accountCount += 1
                            if accountCount % 10 == 0 || eventCount < 50 {
                                print("👤 Account update event")
                            }

                        default:
                            if eventCount < 50 {
                                print("📦 Event: \(messageInfo.type)")
                            }
                        }
                    }

                    // Show periodic statistics
                    if eventCount % 100 == 0 {
                        showStatistics(
                            eventCount: eventCount,
                            commitCount: commitCount,
                            identityCount: identityCount,
                            accountCount: accountCount,
                            postCount: postCount,
                            likeCount: likeCount,
                            followCount: followCount,
                            repostCount: repostCount,
                            profileCount: profileCount,
                            startTime: startTime
                        )
                    }

                case .string(let text):
                    print("📨 Text message: \(text)")

                @unknown default:
                    break
                }

            } catch {
                print("\n⚠️  Connection error: \(error)")
                print("🔄 Attempting to reconnect in 5 seconds...")
                try? await Task.sleep(nanoseconds: 5_000_000_000)

                // Reconnect
                webSocketTask.cancel(with: .goingAway, reason: nil)
                return try await runFirehose()
            }
        }
    }

    static func processOperation(
        action: String,
        path: String,
        postCount: inout Int,
        likeCount: inout Int,
        followCount: inout Int,
        repostCount: inout Int,
        profileCount: inout Int,
        eventCount: Int
    ) {
        if path.contains("app.bsky.feed.post") {
            postCount += 1
            if postCount % 5 == 0 || eventCount < 100 {
                let rkey = path.components(separatedBy: "/").last ?? "unknown"
                print("📝 \(action.uppercased()) post (\(rkey)) | Total: \(postCount)")
            }
        } else if path.contains("app.bsky.feed.like") {
            likeCount += 1
            if likeCount % 10 == 0 {
                print("❤️  \(action.uppercased()) like | Total: \(likeCount)")
            }
        } else if path.contains("app.bsky.graph.follow") {
            followCount += 1
            if followCount % 5 == 0 {
                print("👥 \(action.uppercased()) follow | Total: \(followCount)")
            }
        } else if path.contains("app.bsky.feed.repost") {
            repostCount += 1
            if repostCount % 10 == 0 {
                print("🔄 \(action.uppercased()) repost | Total: \(repostCount)")
            }
        } else if path.contains("app.bsky.actor.profile") {
            profileCount += 1
            if profileCount % 20 == 0 {
                print("👤 \(action.uppercased()) profile | Total: \(profileCount)")
            }
        }
    }

    static func showStatistics(
        eventCount: Int,
        commitCount: Int,
        identityCount: Int,
        accountCount: Int,
        postCount: Int,
        likeCount: Int,
        followCount: Int,
        repostCount: Int,
        profileCount: Int,
        startTime: Date
    ) {
        let elapsed = Date().timeIntervalSince(startTime)
        let rate = Double(eventCount) / max(elapsed, 1.0)

        print("\n" + String(repeating: "=", count: 60))
        print("📊 STATISTICS after \(eventCount) events")
        print(String(repeating: "=", count: 60))
        print("⏱️  Rate: \(String(format: "%.1f", rate)) events/sec")
        print("📦 Events: Commits: \(commitCount) | Identity: \(identityCount) | Account: \(accountCount)")
        print("📝 Posts: \(postCount)")
        print("❤️  Likes: \(likeCount)")
        print("👥 Follows: \(followCount)")
        print("🔄 Reposts: \(repostCount)")
        print("👤 Profiles: \(profileCount)")
        print(String(repeating: "=", count: 60) + "\n")
    }

    // MARK: - CBOR Decoding Helpers

    struct MessageInfo {
        let type: String
        let ops: [(action: String, path: String)]?
    }

    static func decodeFirehoseMessage(_ data: Data) throws -> MessageInfo {
        // Parse message type from CBOR header
        let messageType = parseMessageType(from: data)

        // If it's a commit, also parse operations
        var ops: [(String, String)]? = nil
        if messageType == "#commit" {
            ops = parseRepoOps(from: data)
        }

        return MessageInfo(type: messageType, ops: ops)
    }

    static func parseMessageType(from data: Data) -> String {
        // Look for the "t" field in CBOR which contains the message type
        guard let typeRange = data.range(of: Data([0x74])) else {
            return "unknown"
        }

        let offset = typeRange.upperBound
        guard offset < data.count else { return "unknown" }

        let stringLength = Int(data[offset])
        let stringStart = offset + 1

        guard stringStart + stringLength <= data.count else { return "unknown" }

        let typeData = data[stringStart..<stringStart + stringLength]
        return String(data: typeData, encoding: .utf8) ?? "unknown"
    }

    static func parseRepoOps(from data: Data) -> [(action: String, path: String)] {
        var operations: [(String, String)] = []

        let actionMarker = Data("action".utf8)
        let pathMarker = Data("path".utf8)
        let createMarker = Data("create".utf8)
        let updateMarker = Data("update".utf8)
        let deleteMarker = Data("delete".utf8)

        var searchStart = 0

        while searchStart < data.count - 50 {
            guard let actionRange = data.range(of: actionMarker, in: searchStart..<data.count) else {
                break
            }

            // Determine action type
            var action = "unknown"
            let checkOffset = actionRange.upperBound + 1

            if checkOffset < data.count - 10 {
                let checkData = data[checkOffset..<min(checkOffset + 10, data.count)]
                if checkData.starts(with: createMarker) {
                    action = "create"
                } else if checkData.starts(with: updateMarker) {
                    action = "update"
                } else if checkData.starts(with: deleteMarker) {
                    action = "delete"
                }
            }

            // Find associated path
            let pathSearchEnd = min(actionRange.lowerBound + 300, data.count)
            if let pathRange = data.range(of: pathMarker, in: actionRange.lowerBound..<pathSearchEnd) {
                let pathStart = pathRange.upperBound + 1
                if pathStart < data.count,
                   let pathEnd = data[pathStart...].firstIndex(where: { $0 < 0x20 || $0 > 0x7E }) {
                    let pathData = data[pathStart..<pathEnd]
                    if let path = String(data: pathData, encoding: .utf8), !path.isEmpty {
                        operations.append((action, path))
                    }
                }
            }

            searchStart = actionRange.upperBound
        }

        return operations
    }
}

File: firehose_listener.py
#!/usr/bin/env python3
"""
Bluesky Firehose Listener
Demonstrates unauthenticated connection to the ATProto firehose
"""

import asyncio
import websockets
import cbor2
from datetime import datetime
from collections import defaultdict
import sys


class FirehoseListener:
    def __init__(self, url="wss://bsky.network/xrpc/com.atproto.sync.subscribeRepos"):
        self.url = url
        self.stats = {
            'total_events': 0,
            'commits': 0,
            'identity': 0,
            'account': 0,
            'posts': 0,
            'likes': 0,
            'follows': 0,
            'reposts': 0,
            'profiles': 0,
        }
        self.start_time = datetime.now()

    def decode_message(self, data):
        """Decode a firehose message from DAG-CBOR format"""
        try:
            # The firehose sends two CBOR objects concatenated:
            # 1. Header with 'op' and 't' (type) fields
            # 2. Payload with the actual message data

            from io import BytesIO

            # Decode the header first
            stream = BytesIO(data)
            header = cbor2.load(stream)

            # Get message type
            message_type = header.get('t', 'unknown')
            op = header.get('op', 1)

            if op == -1:
                # Error message
                error = header.get('error', 'Unknown error')
                return {'type': 'error', 'error': error}

            # Try to decode the payload
            try:
                payload = cbor2.load(stream)
            except:
                payload = {}

            return {
                'type': message_type,
                'payload': payload,
                'op': op
            }
        except Exception as e:
            return {'type': 'decode_error', 'error': str(e)}

    def process_commit(self, payload):
        """Process a commit message and extract operations"""
        operations = []

        # Get repo operations if present
        ops = payload.get('ops', [])

        for op in ops:
            action = op.get('action', 'unknown')
            path = op.get('path', '')

            operations.append({
                'action': action,
                'path': path
            })

            # Update statistics
            if 'app.bsky.feed.post' in path:
                self.stats['posts'] += 1
                rkey = path.split('/')[-1] if '/' in path else 'unknown'
                if self.stats['posts'] % 5 == 0 or self.stats['total_events'] < 100:
                    print(f"📝 {action.upper()} post ({rkey}) | Total: {self.stats['posts']}")

            elif 'app.bsky.feed.like' in path:
                self.stats['likes'] += 1
                if self.stats['likes'] % 10 == 0:
                    print(f"❤️  {action.upper()} like | Total: {self.stats['likes']}")

            elif 'app.bsky.graph.follow' in path:
                self.stats['follows'] += 1
                if self.stats['follows'] % 5 == 0:
                    print(f"👥 {action.upper()} follow | Total: {self.stats['follows']}")

            elif 'app.bsky.feed.repost' in path:
                self.stats['reposts'] += 1
                if self.stats['reposts'] % 10 == 0:
                    print(f"🔄 {action.upper()} repost | Total: {self.stats['reposts']}")

            elif 'app.bsky.actor.profile' in path:
                self.stats['profiles'] += 1
                if self.stats['profiles'] % 20 == 0:
                    print(f"👤 {action.upper()} profile | Total: {self.stats['profiles']}")

        return operations

    def show_statistics(self):
        """Display current statistics"""
        elapsed = (datetime.now() - self.start_time).total_seconds()
        rate = self.stats['total_events'] / max(elapsed, 1.0)

        print("\n" + "=" * 60)
        print(f"📊 STATISTICS after {self.stats['total_events']} events")
        print("=" * 60)
        print(f"⏱️  Rate: {rate:.1f} events/sec")
        print(f"📦 Events: Commits: {self.stats['commits']} | Identity: {self.stats['identity']} | Account: {self.stats['account']}")
        print(f"📝 Posts: {self.stats['posts']}")
        print(f"❤️  Likes: {self.stats['likes']}")
        print(f"👥 Follows: {self.stats['follows']}")
        print(f"🔄 Reposts: {self.stats['reposts']}")
        print(f"👤 Profiles: {self.stats['profiles']}")
        print("=" * 60 + "\n")

    async def listen(self):
        """Connect to the firehose and listen for events"""
        print("🔥 Petrel Firehose Listener (Python)")
        print("=" * 60)
        print(f"Connecting to {self.url}")
        print("Press Ctrl+C to stop\n")

        while True:
            try:
                async with websockets.connect(self.url, ping_interval=30) as websocket:
                    print("✅ Connected to firehose!\n")

                    async for message in websocket:
                        self.stats['total_events'] += 1

                        # Decode the message
                        decoded = self.decode_message(message)

                        message_type = decoded.get('type', 'unknown')

                        if message_type == '#commit':
                            self.stats['commits'] += 1
                            payload = decoded.get('payload', {})
                            self.process_commit(payload)

                        elif message_type == '#identity':
                            self.stats['identity'] += 1
                            if self.stats['identity'] % 10 == 0 or self.stats['total_events'] < 50:
                                print("🆔 Identity update event")

                        elif message_type == '#account':
                            self.stats['account'] += 1
                            if self.stats['account'] % 10 == 0 or self.stats['total_events'] < 50:
                                print("👤 Account update event")

                        elif message_type == 'error':
                            print(f"❌ Error from server: {decoded.get('error')}")

                        elif message_type == 'decode_error':
                            if self.stats['total_events'] < 10:
                                print(f"⚠️  Decode error: {decoded.get('error')}")

                        # Show periodic statistics
                        if self.stats['total_events'] % 100 == 0:
                            self.show_statistics()

            except websockets.exceptions.ConnectionClosed:
                print("\n⚠️  Connection closed. Reconnecting in 5 seconds...")
                await asyncio.sleep(5)

            except KeyboardInterrupt:
                print("\n\n👋 Shutting down firehose listener...")
                break

            except Exception as e:
                print(f"\n❌ Error: {e}")
                print("🔄 Reconnecting in 5 seconds...")
                await asyncio.sleep(5)


async def main():
    listener = FirehoseListener()
    await listener.listen()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n👋 Goodbye!")
        sys.exit(0)
